lib/rbac.ex

```diff
@@ -37,4 +37,117 @@ defmodule RBAC do
+  `get_approles/2` fetches the roles for the app
+  """
+  def get_approles(auth_url, client_id) do
+    url = "#{auth_url}/approles/#{client_id}"
+    HTTPoison.start()
+
+    HTTPoison.get(url)
+    |> parse_body_response()
+  end
+
+  @doc """
+  `parse_body_response/1` parses the response
+  so your app can use the resulting JSON (list of roles).
+  """
+  @spec parse_body_response({atom, String.t()}) :: String.t()
+  def parse_body_response({:error, err}), do: {:error, err}
+
+  def parse_body_response({:ok, response}) do
+    body = Map.get(response, :body)
+    # IO.inspect(body)
+    # make keys of map atoms for easier access in templates
+    if body == nil do
+      {:error, :no_body}
+    else
+      {:ok, str_key_map} = Jason.decode(body)
+
+      atom_key_map =
+        Enum.map(str_key_map, fn role ->
+          for {key, val} <- role, into: %{}, do: {String.to_atom(key), val}
+        end)
+
+      {:ok, atom_key_map}
+    end
+
+    # https://stackoverflow.com/questions/31990134
+  end
+
+  @doc """
+  `init_roles/2 fetches the list of roles for an app 
+  from the auth app (auth_url) based on the client_id
+  and caches the list for fast access. 
+  ETS is an in-memory cache you get for *Free* in Elixir/Erlang.
+  See: https://elixir-lang.org/getting-started/mix-otp/ets.html
+  and: https://elixirschool.com/en/lessons/specifics/ets
+  """
+  def init_roles_cache(auth_url, client_id) do
+    {:ok, roles} = RBAC.get_approles(auth_url, client_id)
+    :ets.new(:roles_cache, [:set, :protected, :named_table])
+    # insert full list:
+    :ets.insert(:roles_cache, {"roles", roles})
+    # insert individual roles for fast lookup:
+    Enum.each(roles, fn role ->
+      :ets.insert(:roles_cache, {role.name, role})
+      :ets.insert(:roles_cache, {role.id, role})
+    end)
+  end
+
+  @doc """
+  `get_role_from_cache/1 retrieves a role from ets cache
+  """
+  def get_role_from_cache(term) do
+    case :ets.lookup(:roles_cache, term) do
+      # not found:
+      [] -> %{id: 0}
+      # extract role:
+      [{_term, role}] -> role
+    end
+  end
+
+  @doc """
+  `has_role?/2 confirms if the person has the given role
+  e.g: 
+  has_role?(conn, "home_admin") > true
+  has_role?(conn, "potus") > false
+  """
+  def has_role?(conn, role_name) do
+    role = get_role_from_cache(role_name)
+
+    person_roles =
+      conn.assigns.person.roles
+      |> String.split(",", trim: true)
+      |> Enum.map(&String.to_integer/1)
+
+    Enum.member?(person_roles, role.id)
+  end
+
+  @doc """
+  `has_role_any/2 checks if the person has any one (or more)
+  of the roles listed. Allows multiple roles to access content.
+  e.g: 
+  has_role_any?(conn, ["home_admin", "building_owner") > true
+  has_role_any?(conn, ["potus", "el_presidente") > false
+  """
+  def has_role_any?(conn, roles_list) do
+    list_ids = Enum.map(roles_list, fn role -> 
+      r = get_role_from_cache(role)
+      r.id
+    end)
+
+    # list of integers
+    person_roles =
```

Comment:
DRY - Getting `people_roles` should be moved into a separate function

Reply:
Yeah, agreed. Can easily be moved out. I can do this now (_unless you are already doing some refactoring ..._) 💭

Reply:
Created DRY function: 
https://github.com/dwyl/rbac/blob/1ea682cc39be950a0c54ed6f2eef3d59da603807/lib/rbac.ex#L119-L125